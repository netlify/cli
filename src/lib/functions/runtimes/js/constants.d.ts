export declare const SECONDS_TO_MILLISECONDS = 1000;
//# sourceMappingURL=constants.d.ts.map