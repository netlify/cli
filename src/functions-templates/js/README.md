## note to devs

place new templates here and our CLI will pick it up. each template must be in its own folder.

## not a long term solution

we dont want people to update their CLI every time we add a template. see
https://github.com/netlify/netlify-dev-plugin/issues/42 for how we may solve in future

## template lifecycles

- onComplete
  - meant for messages, logging, light cleanup
- onAllAddonsInstalled?
  - not implemented yet
  - meant for heavier work, but not sure if different from onComplete

## template addons

specify an array of objects of this shape:

```ts
{
  addonName: String,
  addonDidInstall?: Function // for executing arbitrary postinstall code for a SINGLE addon
}
```

## why place templates in a separate folder

we dont colocate this inside `src/commands/functions` because oclif will think it's a new command.

every function should be registered with their respective `template-registry.js`.

## typescript and go

we have some templates here but they are unused for now until Netlify Dev supports them.
