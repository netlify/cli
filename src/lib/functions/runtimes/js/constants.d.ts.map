{"version":3,"file":"constants.d.ts","sourceRoot":"","sources":["constants.ts"],"names":[],"mappings":"AAAA,eAAO,MAAM,uBAAuB,OAAO,CAAA"}