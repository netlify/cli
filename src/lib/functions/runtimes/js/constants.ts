export const SECONDS_TO_MILLISECONDS = 1000
