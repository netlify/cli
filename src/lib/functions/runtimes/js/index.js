const { dirname } = require('path')

const lambdaLocal = require('lambda-local')
const winston = require('winston')

const detectNetlifyLambdaBuilder = require('./builders/netlify-lambda')
const detectZisiBuilder = require('./builders/zisi')

const SECONDS_TO_MILLISECONDS = 1e3

let netlifyLambdaDetectorCache

const logger = winston.createLogger({
  levels: winston.config.npm.levels,
  transports: [new winston.transports.Console({ level: 'warn' })],
})

lambdaLocal.setLogger(logger)

// The netlify-lambda builder can't be enabled or disabled on a per-function
// basis and its detection mechanism is also quite expensive, so we detect
// it once and cache the result.
const detectNetlifyLambdaWithCache = () => {
  if (netlifyLambdaDetectorCache === undefined) {
    netlifyLambdaDetectorCache = detectNetlifyLambdaBuilder()
  }

  return netlifyLambdaDetectorCache
}

const getBuildFunction = async ({ config, directory, errorExit, func, projectRoot }) => {
  const netlifyLambdaBuilder = await detectNetlifyLambdaWithCache()

  if (netlifyLambdaBuilder) {
    return netlifyLambdaBuilder.build
  }

  const zisiBuilder = await detectZisiBuilder({ config, directory, errorExit, func, projectRoot })

  if (zisiBuilder) {
    return zisiBuilder.build
  }

  // If there's no function builder, we create a simple one on-the-fly which
  // returns as `srcFiles` the function directory, if there is one, or its
  // main file otherwise.
  const functionDirectory = dirname(func.mainFile)
  const srcFiles = functionDirectory === directory ? [func.mainFile] : [functionDirectory]
  const schedule = await detectZisiBuilder.parseForSchedule({ mainFile: func.mainFile, config, projectRoot })

  return () => ({ schedule, srcFiles })
}

const invokeFunction = async ({ context, event, func, timeout }) => {
  // If a function builder has defined a `buildPath` property, we use it.
  // Otherwise, we'll invoke the function's main file.
  const lambdaPath = (func.buildData && func.buildData.buildPath) || func.mainFile
  const result = await lambdaLocal.execute({
    clientContext: JSON.stringify(context),
    event,
    lambdaPath,
    timeoutMs: timeout * SECONDS_TO_MILLISECONDS,
    verboseLevel: 3,
  })

  return result
}

const onDirectoryScan = async () => {
  const netlifyLambdaBuilder = await detectNetlifyLambdaWithCache()

  // Before we start a directory scan, we check whether netlify-lambda is being
  // used. If it is, we run it, so that the functions directory is populated
  // with the compiled files before the scan begins.
  if (netlifyLambdaBuilder) {
    await netlifyLambdaBuilder.build()
  }
}

module.exports = { getBuildFunction, invokeFunction, name: 'js', onDirectoryScan }
