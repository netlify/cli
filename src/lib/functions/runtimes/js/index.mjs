import { dirname } from 'path'

import lambdaLocal from '@skn0tt/lambda-local'
import winston from 'winston'

import detectNetlifyLambdaBuilder from './builders/netlify-lambda.mjs'
import detectZisiBuilder, { parseForSchedule } from './builders/zisi.mjs'

export const name = 'js'

const SECONDS_TO_MILLISECONDS = 1000

let netlifyLambdaDetectorCache

const logger = winston.createLogger({
  levels: winston.config.npm.levels,
  transports: [new winston.transports.Console({ level: 'warn' })],
})

lambdaLocal.setLogger(logger)

// The netlify-lambda builder can't be enabled or disabled on a per-function
// basis and its detection mechanism is also quite expensive, so we detect
// it once and cache the result.
const detectNetlifyLambdaWithCache = () => {
  if (netlifyLambdaDetectorCache === undefined) {
    netlifyLambdaDetectorCache = detectNetlifyLambdaBuilder()
  }

  return netlifyLambdaDetectorCache
}

export const getBuildFunction = async ({ config, directory, errorExit, func, projectRoot }) => {
  const netlifyLambdaBuilder = await detectNetlifyLambdaWithCache()

  if (netlifyLambdaBuilder) {
    return netlifyLambdaBuilder.build
  }

  const zisiBuilder = await detectZisiBuilder({ config, directory, errorExit, func, projectRoot })

  if (zisiBuilder) {
    return zisiBuilder.build
  }

  // If there's no function builder, we create a simple one on-the-fly which
  // returns as `srcFiles` the function directory, if there is one, or its
  // main file otherwise.
  const functionDirectory = dirname(func.mainFile)
  const srcFiles = functionDirectory === directory ? [func.mainFile] : [functionDirectory]
  const schedule = await parseForSchedule({ mainFile: func.mainFile, config, projectRoot })

  return () => ({ schedule, srcFiles })
}

export const invokeFunction = async ({ context, event, func, timeout }) => {
  // If a function builder has defined a `buildPath` property, we use it.
  // Otherwise, we'll invoke the function's main file.
  const lambdaPath = (func.buildData && func.buildData.buildPath) || func.mainFile
  const result = await lambdaLocal.execute({
    clientContext: JSON.stringify(context),
    event,
    lambdaPath,
    timeoutMs: timeout * SECONDS_TO_MILLISECONDS,
    verboseLevel: 3,
  })

  return result
}

export const onDirectoryScan = async () => {
  const netlifyLambdaBuilder = await detectNetlifyLambdaWithCache()

  // Before we start a directory scan, we check whether netlify-lambda is being
  // used. If it is, we run it, so that the functions directory is populated
  // with the compiled files before the scan begins.
  if (netlifyLambdaBuilder) {
    await netlifyLambdaBuilder.build()
  }
}
