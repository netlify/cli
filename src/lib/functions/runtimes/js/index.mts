// @ts-expect-error TS(2451): Cannot redeclare block-scoped variable 'dirname'.
const { dirname } = require('path')

const lambdaLocal = require('lambda-local')
const winston = require('winston')

const detectNetlifyLambdaBuilder = require('./builders/netlify-lambda.cjs')
const detectZisiBuilder = require('./builders/zisi.cjs')

// @ts-expect-error TS(2451): Cannot redeclare block-scoped variable 'SECONDS_TO... Remove this comment to see the full error message
const SECONDS_TO_MILLISECONDS = 1e3

// @ts-expect-error TS(2304): Cannot find name '$TSFixMe'.
let netlifyLambdaDetectorCache: $TSFixMe

const logger = winston.createLogger({
  levels: winston.config.npm.levels,
  transports: [new winston.transports.Console({ level: 'warn' })],
})

lambdaLocal.setLogger(logger)

// The netlify-lambda builder can't be enabled or disabled on a per-function
// basis and its detection mechanism is also quite expensive, so we detect
// it once and cache the result.
const detectNetlifyLambdaWithCache = () => {
  if (netlifyLambdaDetectorCache === undefined) {
    netlifyLambdaDetectorCache = detectNetlifyLambdaBuilder()
  }

  return netlifyLambdaDetectorCache
}

// @ts-expect-error TS(2451): Cannot redeclare block-scoped variable 'getBuildFu... Remove this comment to see the full error message
const getBuildFunction = async ({
  config,
  directory,
  errorExit,
  func,
  projectRoot
// @ts-expect-error TS(2304): Cannot find name '$TSFixMe'.
}: $TSFixMe) => {
  const netlifyLambdaBuilder = await detectNetlifyLambdaWithCache()

  if (netlifyLambdaBuilder) {
    return netlifyLambdaBuilder.build
  }

  const zisiBuilder = await detectZisiBuilder({ config, directory, errorExit, func, projectRoot })

  if (zisiBuilder) {
    return zisiBuilder.build
  }

  // If there's no function builder, we create a simple one on-the-fly which
  // returns as `srcFiles` the function directory, if there is one, or its
  // main file otherwise.
  const functionDirectory = dirname(func.mainFile)
  const srcFiles = functionDirectory === directory ? [func.mainFile] : [functionDirectory]
  const schedule = await detectZisiBuilder.parseForSchedule({ mainFile: func.mainFile, config, projectRoot })

  return () => ({ schedule, srcFiles })
}

// @ts-expect-error TS(2451): Cannot redeclare block-scoped variable 'invokeFunc... Remove this comment to see the full error message
const invokeFunction = async ({
  context,
  event,
  func,
  timeout
// @ts-expect-error TS(2304): Cannot find name '$TSFixMe'.
}: $TSFixMe) => {
  // If a function builder has defined a `buildPath` property, we use it.
  // Otherwise, we'll invoke the function's main file.
  const lambdaPath = (func.buildData && func.buildData.buildPath) || func.mainFile
  const result = await lambdaLocal.execute({
    clientContext: JSON.stringify(context),
    event,
    lambdaPath,
    timeoutMs: timeout * SECONDS_TO_MILLISECONDS,
    verboseLevel: 3,
  })

  return result
}

const onDirectoryScan = async () => {
  const netlifyLambdaBuilder = await detectNetlifyLambdaWithCache()

  // Before we start a directory scan, we check whether netlify-lambda is being
  // used. If it is, we run it, so that the functions directory is populated
  // with the compiled files before the scan begins.
  if (netlifyLambdaBuilder) {
    await netlifyLambdaBuilder.build()
  }
}

module.exports = { getBuildFunction, invokeFunction, name: 'js', onDirectoryScan }
