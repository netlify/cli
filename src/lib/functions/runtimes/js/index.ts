import { createConnection } from 'net'
import { dirname } from 'path'
import { pathToFileURL } from 'url'
import { Worker } from 'worker_threads'

import lambdaLocal, { type LambdaEvent } from 'lambda-local'

import type { BuildFunction, GetBuildFunction, InvokeFunction } from '../index.js'
import { BLOBS_CONTEXT_VARIABLE } from '../../../blobs/blobs.js'
import type NetlifyFunction from '../../netlify-function.js'

import detectZisiBuilder, { getFunctionMetadata, ZisiBuildResult } from './builders/zisi.js'
import { SECONDS_TO_MILLISECONDS } from './constants.js'
import type { WorkerMessage } from './worker.js'

export const name = 'js'

type SimpleJsBuildResult = {
  schedule?: string
  srcFiles: string[]
}

export type JsBuildResult = ZisiBuildResult | SimpleJsBuildResult

// TODO(serhalp): Unify these. This is bonkers that the two underlying invocation mechanisms are encapsulated but we
// return slightly different shapes for them.
export type JsInvokeFunctionResult = WorkerMessage | LambdaEvent

lambdaLocal.getLogger().level = 'alert'

export async function getBuildFunction({
  config,
  directory,
  errorExit,
  func,
  projectRoot,
}: Parameters<GetBuildFunction<JsBuildResult>>[0]) {
  const metadata = await getFunctionMetadata({ mainFile: func.mainFile, config, projectRoot })
  const zisiBuilder = await detectZisiBuilder({ config, directory, errorExit, func, metadata, projectRoot })

  if (zisiBuilder) {
    return zisiBuilder.build
  }

  // If there's no function builder, we create a simple one on-the-fly which
  // returns as `srcFiles` the function directory, if there is one, or its
  // main file otherwise.
  const functionDirectory = dirname(func.mainFile)
  const srcFiles = functionDirectory === directory ? [func.mainFile] : [functionDirectory]

  const build: BuildFunction<JsBuildResult> = () => Promise.resolve({ schedule: metadata?.schedule, srcFiles })
  return build
}

const workerURL = new URL('worker.js', import.meta.url)

export const invokeFunction = async ({
  context,
  environment,
  event,
  func,
  timeout,
}: Parameters<InvokeFunction<JsBuildResult>>[0]): Promise<JsInvokeFunctionResult> => {
  const { buildData } = func
  // I have no idea why, but it appears that treating the case of a missing `buildData` or missing
  // `buildData.runtimeAPIVersion` as V1 is important.
  const runtimeAPIVersion =
    buildData != null && 'runtimeAPIVersion' in buildData && typeof buildData.runtimeAPIVersion === 'number'
      ? buildData.runtimeAPIVersion
      : null
  if (runtimeAPIVersion == null || runtimeAPIVersion !== 2) {
    return await invokeFunctionDirectly({
      context,
      environment: environment as Record<string, string>,
      event,
      func,
      timeout,
    })
  }

  const workerData = {
    clientContext: JSON.stringify(context),
    environment,
    event,
    // If a function builder has defined a `buildPath` property, we use it.
    // Otherwise, we'll invoke the function's main file.
    // Because we use import() we have to use file:// URLs for Windows.
    entryFilePath: pathToFileURL(
      buildData != null && 'buildPath' in buildData && buildData.buildPath ? buildData.buildPath : func.mainFile,
    ).href,
    timeoutMs: timeout * SECONDS_TO_MILLISECONDS,
  }

  const worker = new Worker(workerURL, { workerData })
  return await new Promise((resolve, reject) => {
    worker.on('message', (result: WorkerMessage): void => {
      // TODO(serhalp): Improve `WorkerMessage` type. It sure would be nice to keep it simple as it
      // is now, but technically this is an arbitrary type from the user function return...
      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
      if (result?.streamPort != null) {
        const client = createConnection(
          {
            port: result.streamPort,
            host: 'localhost',
          },
          () => {
            result.body = client
            resolve(result)
          },
        )
        client.on('error', reject)
      } else {
        resolve(result)
      }
    })

    worker.on('error', reject)
  })
}

export const invokeFunctionDirectly = async <BuildResult extends JsBuildResult>({
  context,
  environment,
  event,
  func,
  timeout,
}: {
  context: Record<string, unknown>
  environment: Record<string, string>
  event: Record<string, unknown>
  func: NetlifyFunction<BuildResult>
  timeout: number
}): Promise<LambdaEvent> => {
  const buildData = await func.getBuildData()
  if (buildData == null) {
    throw new Error('Cannot invoke a function that has not been built')
  }
  // If a function builder has defined a `buildPath` property, we use it.
  // Otherwise, we'll invoke the function's main file.
  const lambdaPath =
    'buildPath' in buildData && typeof buildData.buildPath === 'string' ? buildData.buildPath : func.mainFile
  const result = await lambdaLocal.execute({
    clientContext: JSON.stringify(context),
    environment: {
      // Include environment variables from config
      ...environment,
      // We've set the Blobs context on the parent process, which means it will
      // be available to the Lambda. This would be inconsistent with production
      // where only V2 functions get the context injected. To fix it, unset the
      // context variable before invoking the function.
      // This has the side-effect of also removing the variable from `process.env`.
      [BLOBS_CONTEXT_VARIABLE]: undefined,
    },
    event,
    lambdaPath,
    timeoutMs: timeout * SECONDS_TO_MILLISECONDS,
    verboseLevel: 3,
    esm: lambdaPath.endsWith('.mjs'),
  })

  return result
}
