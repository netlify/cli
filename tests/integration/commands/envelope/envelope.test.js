import { describe, test } from 'vitest'

import { callCli } from '../../utils/call-cli.js'
import { getCLIOptions, withMockApi } from '../../utils/mock-api.js'
import { withSiteBuilder } from '../../utils/site-builder.ts'
import { normalize } from '../../utils/snapshots.js'

const siteInfo = {
  account_slug: 'test-account',
  build_settings: {
    env: {},
  },
  id: 'site_id',
  name: 'site-name',
}
const existingVar = {
  key: 'EXISTING_VAR',
  scopes: ['builds', 'functions'],
  values: [
    {
      id: '1234',
      context: 'production',
      value: 'envelope-prod-value',
    },
    {
      id: '2345',
      context: 'dev',
      value: 'envelope-dev-value',
    },
  ],
}
const otherVar = {
  key: 'OTHER_VAR',
  scopes: ['builds', 'functions', 'runtime', 'post_processing'],
  values: [
    {
      id: '3456',
      context: 'all',
      value: 'envelope-all-value',
    },
  ],
}
const response = [existingVar, otherVar]
const routes = [
  { path: 'sites/site_id', response: siteInfo },
  { path: 'sites/site_id/service-instances', response: [] },
  {
    path: 'accounts',
    response: [{ slug: siteInfo.account_slug }],
  },
  {
    path: 'accounts/test-account/env/EXISTING_VAR',
    response: existingVar,
  },
  {
    path: 'accounts/test-account/env/OTHER_VAR',
    response: otherVar,
  },
  {
    path: 'accounts/test-account/env',
    response,
  },
  {
    path: 'accounts/test-account/env',
    method: 'POST',
    response: {},
  },
  {
    path: 'accounts/test-account/env/EXISTING_VAR',
    method: 'PUT',
    response: {},
  },
  {
    path: 'accounts/test-account/env/EXISTING_VAR',
    method: 'PATCH',
    response: {},
  },
  {
    path: 'accounts/test-account/env/EXISTING_VAR',
    method: 'DELETE',
    response: {},
  },
  {
    path: 'accounts/test-account/env/EXISTING_VAR/value/1234',
    method: 'DELETE',
    response: {},
  },
  {
    path: 'accounts/test-account/env/OTHER_VAR',
    method: 'PATCH',
    response: {},
  },
  {
    path: 'accounts/test-account/env/OTHER_VAR',
    method: 'PUT',
    response: {},
  },
  {
    path: 'accounts/test-account/env/OTHER_VAR',
    method: 'DELETE',
    response: {},
  },
  {
    path: 'accounts/test-account/env/OTHER_VAR/value/3456',
    method: 'DELETE',
    response: {},
  },
]

describe.concurrent('command/envelope', () => {
  test('env:import should throw error if file not exists', async (t) => {
    await withSiteBuilder(t, async (builder) => {
      await builder.build()

      await withMockApi(routes, async ({ apiUrl }) => {
        t.expect(callCli(['env:import', '.env'], getCLIOptions({ builder, apiUrl }))).rejects.toThrow()
      })
    })
  })

  test('env:import --json should import new vars and override existing vars', async (t) => {
    await withSiteBuilder(t, async (builder) => {
      const finalEnv = {
        EXISTING_VAR: 'from-dotenv',
        OTHER_VAR: 'envelope-all-value',
        NEW_VAR: 'from-dotenv',
      }

      await builder
        .withEnvFile({
          path: '.env',
          env: {
            EXISTING_VAR: 'from-dotenv',
            NEW_VAR: 'from-dotenv',
          },
        })
        .build()

      await withMockApi(routes, async ({ apiUrl }) => {
        const cliResponse = await callCli(['env:import', '--json', '.env'], getCLIOptions({ builder, apiUrl }), true)

        t.expect(cliResponse).toStrictEqual(finalEnv)
      })
    })
  })

  test('env:import --json --replace-existing should replace all existing vars and return imported', async (t) => {
    await withSiteBuilder(t, async (builder) => {
      const finalEnv = {
        EXISTING_VAR: 'from-dotenv',
        NEW_VAR: 'from-dotenv',
      }

      await builder
        .withEnvFile({
          path: '.env',
          env: {
            EXISTING_VAR: 'from-dotenv',
            NEW_VAR: 'from-dotenv',
          },
        })
        .build()

      await withMockApi(routes, async ({ apiUrl }) => {
        const cliResponse = await callCli(
          ['env:import', '--replace-existing', '--json', '.env'],
          getCLIOptions({ builder, apiUrl }),
          true,
        )

        t.expect(cliResponse).toStrictEqual(finalEnv)
      })
    })
  })

  test('env:clone should return success message', async (t) => {
    const siteInfoFrom = {
      ...siteInfo,
      id: 'site_id_a',
      name: 'site-name-a',
    }

    const siteInfoTo = {
      ...siteInfo,
      id: 'site_id_b',
      name: 'site-name-b',
    }

    const cloneRoutes = [
      { path: 'sites/site_id', response: siteInfo },
      { path: 'sites/site_id_a', response: siteInfoFrom },
      { path: 'sites/site_id_b', response: siteInfoTo },
      { path: 'sites/site_id/service-instances', response: [] },
      {
        path: 'accounts',
        response: [{ slug: siteInfo.account_slug }],
      },
      {
        path: 'accounts/test-account/env',
        response,
      },
      {
        path: 'accounts/test-account/env',
        method: 'POST',
        response: {},
      },
      {
        path: 'accounts/test-account/env/EXISTING_VAR',
        method: 'DELETE',
        response: {},
      },
      {
        path: 'accounts/test-account/env/OTHER_VAR',
        method: 'DELETE',
        response: {},
      },
    ]

    await withSiteBuilder(t, async (builder) => {
      await builder.build()
      await withMockApi(cloneRoutes, async ({ apiUrl, requests }) => {
        const cliResponse = await callCli(
          ['env:clone', '--from', 'site_id_a', '--to', 'site_id_b'],
          getCLIOptions({ apiUrl, builder }),
        )

        t.expect(normalize(cliResponse)).toMatchSnapshot()

        const deleteRequests = requests.filter((request) => request.method === 'DELETE')
        t.expect(deleteRequests.length).toBe(2)

        const postRequest = requests.find((request) => request.method === 'POST')
        t.expect(postRequest.body.map(({ key }) => key)).toStrictEqual(['EXISTING_VAR', 'OTHER_VAR'])
      })
    })
  })
})
