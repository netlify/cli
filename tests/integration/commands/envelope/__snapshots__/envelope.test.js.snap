// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`command/envelope > env:clone should return success message 1`] = `"Successfully cloned environment variables from site-name-a to site-name-b"`;
